org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/selectionRange/LSPSelectionRangeAbstractHandler.java

```diff
@@ -0,0 +1,218 @@
+
+		private SelectionRange root;
+
+		private SelectionRange previous;
+
+		private final StyledText styledText;
+
+		private boolean updating;
+
+		public void setRoot(SelectionRange root) {
+			this.root = root;
+			this.previous = root;
+		}
+
+		public static SelectionRangeHandler getSelectionRangeHandler(StyledText styledText) {
+			SelectionRangeHandler handler = (SelectionRangeHandler) styledText.getData(KEY);
+			if (handler == null) {
+				handler = new SelectionRangeHandler(styledText);
+			}
+			return handler;
+		}
+
+		public SelectionRangeHandler(StyledText styledText) {
+			this.styledText = styledText;
+			styledText.setData(KEY, this);
+			styledText.addCaretListener(new CaretListener() {
+
+				@Override
+				public void caretMoved(CaretEvent arg0) {
+					if (!updating) {
+						root = null;
+					}
+				}
+			});
+		}
+
+		public SelectionRange getSelectionRange(Direction direction) {
+			if (direction == Direction.UP) {
+				if (previous != null) {
+					previous = previous.getParent();
+					return previous;
+				}
+			} else {
+				if (previous != null) {
+					SelectionRange selectionRange = root;
+					while (selectionRange != null) {
+						SelectionRange parent = selectionRange.getParent();
+						if (previous.equals(parent)) {
+							previous = selectionRange;
+							return previous;
+						}
+						selectionRange = parent;
+					}
+				}
+			}
+			return null;
+		}
+
+		public boolean isDirty() {
+			return root == null;
+		}
+
+		public void updateSelection(ISelectionProvider provider, IDocument document, Direction direction) {
+			if (styledText.isDisposed()) {
+				return;
+			}
+			SelectionRange selectionRange = getSelectionRange(direction);
+			if (selectionRange != null) {
+				ISelection selection = LSPEclipseUtils.toSelection(selectionRange.getRange(), document);
+				styledText.getDisplay().execute(() -> {
+					try {
+						updating = true;
+						provider.setSelection(selection);
+					} finally {
+						updating = false;
+					}
+				});
+			}
+		}
+
+	}
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException {
+		IEditorPart part = HandlerUtil.getActiveEditor(event);
+		if (part instanceof ITextEditor textEditor) {
+			final ISelectionProvider provider = textEditor.getSelectionProvider();
+			if (provider == null) {
+				return null;
+			}
+			ISelection sel = provider.getSelection();
+			ITextViewer viewer = textEditor.getAdapter(ITextViewer.class);
+			if (viewer == null) {
+				return null;
+			}
+			StyledText styledText = viewer.getTextWidget();
+			if (styledText == null) {
+				return null;
+			}
+
+			if (sel instanceof ITextSelection textSelection && !textSelection.isEmpty()) {
+				IDocument document = LSPEclipseUtils.getDocument(textEditor);
+				if (document != null) {
+					LanguageServerDocumentExecutor executor = LanguageServers.forDocument(document)
+							.withCapability(ServerCapabilities::getSelectionRangeProvider);
+					if (executor.anyMatching()) {

```

Comment:
I did that because I wanted to avoid storing in Styledtext the SelectionRangeHandler instance if there are no LS which supports SelectionRange but I can change that if you prefer.

Reply:
I am not sure I understand that, so I do not know if it is better to change it or, but at least we should add a comment why it is like it is at the moment, I think it is not something one can figure out from the code. What do you think?

Reply:
@rubenporras I try to add some comments, please tell me if it better and more understandable.

Reply:
Yes, it is a bit better now. I understand what you mean in your last post.

BTW: Do you think it would be possible to have some test?

Reply:
I fear that I will not have time to do that now -(